### PR TITLE
Add AnsibleVarsFrom

### PR DESCRIPTION
**Pull request by fao89**
Inspired by Pod's Env x EnvFrom:
https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/#define-an-environment-variable-for-a-container

https://github.com/kubernetes/kubernetes/blob/v1.28.7/pkg/apis/core/types.go#L2321-L2330

https://github.com/kubernetes/kubernetes/blob/v1.28.7/pkg/kubelet/kubelet_pods.go#L619

AnsibleVarsFrom will add variables from ConfigMap/Secret to the inventory,  variables from AnsibleVars will take precedence

### CODE REVIEW

**Review by slagle — COMMENTED** [review]
+2

Please add a corresponding docs patch that shows how this would be used with `edpm_bootstrap_command`.

File: pkg/deployment/inventory.go [hunk 1]
```diff
@@ -34,13 +37,91 @@ import (
...
+		case varFrom.ConfigMapRef != nil:
+			cm := varFrom.ConfigMapRef
+			name := cm.Name
+			configMap, ok := configMaps[name]
+			if !ok {
+				optional := cm.Optional != nil && *cm.Optional
+				configMap = &v1.ConfigMap{}
+				err := client.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, configMap)
+				if err != nil {
+					if errors.IsNotFound(err) && optional {
```
Comment (jpodivin) [inline pkg/deployment/inventory.go:63]: I would prefer if this code raised a warning  about the config map missing.

File: pkg/deployment/inventory.go [hunk 2]
```diff
@@ -34,13 +37,91 @@ import (
...
+		case varFrom.SecretRef != nil:
+			s := varFrom.SecretRef
+			name := s.Name
+			secret, ok := secrets[name]
+			if !ok {
+				optional := s.Optional != nil && *s.Optional
+				secret = &v1.Secret{}
+				err := client.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, secret)
+				if err != nil {
+					if errors.IsNotFound(err) && optional {
```
Comment (jpodivin) [inline pkg/deployment/inventory.go:89]: ditto

File: pkg/deployment/inventory.go [hunk 3]
```diff
@@ -34,13 +37,91 @@ import (
...
 // GenerateNodeSetInventory yields a parsed Inventory for role
 func GenerateNodeSetInventory(ctx context.Context, helper *helper.Helper,
 	instance *dataplanev1.OpenStackDataPlaneNodeSet,
 	allIPSets map[string]infranetworkv1.IPSet, dnsAddresses []string, defaultImages dataplanev1.DataplaneAnsibleImageDefaults) (string, error) {
 	inventory := ansible.MakeInventory()
 	nodeSetGroup := inventory.AddGroup(instance.Name)
-	err := resolveGroupAnsibleVars(&instance.Spec.NodeTemplate, &nodeSetGroup, defaultImages)
+	groupVars, err := getAnsibleVarsFrom(ctx, helper, instance.Namespace, &instance.Spec.NodeTemplate.Ansible)
+	if err != nil {
+		utils.LogErrorForObject(helper, err, "Could not resolve ansible group vars", instance)
```
Comment (jpodivin) [inline pkg/deployment/inventory.go:118]: This error duplicates the one on line 126. However, it has a different cause. It should indicate that it was related to the retrieval of vars from config map .

File: Makefile [hunk 4]
```diff
@@ -98,7 +98,8 @@ docs: manifests docs-dependencies crd-to-markdown ## Build docs
 	$(CRD_MARKDOWN) -f api/v1beta1/common.go -f api/v1beta1/openstackdataplaneservice_types.go -f api/v1beta1/openstackdataplanenodeset_types.go -f api/v1beta1/openstackdataplanedeployment_types.go -n OpenStackDataPlaneService -n OpenStackDataPlaneNodeSet -n OpenStackDataPlaneDeployment > docs/assemblies/custom_resources.md
 	bundle exec kramdoc --auto-ids docs/assemblies/custom_resources.md && rm docs/assemblies/custom_resources.md
 	sed -i "s/=== Custom/== Custom/g" docs/assemblies/custom_resources.adoc
-	cd docs; $(MAKE) html
+	cd docs; $(MAKE) html BUILD=upstream
```
Comment (jpodivin) [inline Makefile:101]: I would put this change into it's own PR if possible. It isn't directly related to the feature proposed.

Comment (fao89) [reply Makefile:101]: I need this on this PR so the CI can render the downstream version of the docs: https://logserver.rdoproject.org/23/723/6960badc997091a6fd3ac12d0a82d322477d8249/github-check/dataplane-operator-docs-preview/8be5764/docs_build/docs_build/dataplane/

File: examples/base/vars/nic-config.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: network-config-template
+data:
+  network_config_template: |
```
Comment (fao89) [inline examples/base/vars/nic-config.yaml:6]: reusing the `edpm_network_config_template` var

File: examples/base/crs/openstackdataplanenodeset.yaml [hunk 6]
```diff
@@ -44,54 +44,27 @@ spec:
   nodeTemplate:
     ansibleSSHPrivateKeySecret: dataplane-ansible-ssh-private-key-secret
     ansible:
+      ansibleVarsFrom:
+        - prefix: edpm_
+          configMapRef:
+            name: network-config-template
```
Comment (fao89) [inline examples/base/crs/openstackdataplanenodeset.yaml:50]: reusing the `edpm_network_config_template` var

File: Makefile (hunk 4 above)
Comment (jpodivin) [reply Makefile:101]: Ok. If there is no other way to address the issue.

**Review by jpodivin — APPROVED** [review]
